readme.md

```diff
@@ -15,834 +14,1113 @@
+		editor.setCurrentSelection( that );
+	} );
+}
 
-    var id = 0,
-        elements = [],
-        matches = query( '#foo' ),
-        length = matches.length;
+function someCallback() {
+	// Should refers to Selection, not Editor.
+	this;
+}
+```
 
-    for ( ; id < length; id++ )
-        elements.push( matches[ id ] );
+### 2.7.2. Use `thisArg` argument
 
-    ```
+Several prototype methods of ES 5.1 built-ins come with a special `thisArg` signature, which should be used whenever possible. CKEditor follows this practice too, so for example [`eventTarget.on`](http://docs.ckeditor.com/#!/api/CKEDITOR.event-method-on) also accepts context as an argument.
 
-    A few additional naming pointers:
+```javascript
+var obj = { f: 'foo', b: 'bar', q: 'qux' };
 
-    ```javascript
+Object.keys( obj ).forEach( function( key ) {
 
-    // 6.A.3.1
-    // Naming strings
+  // |this| now refers to `obj`
 
-    `dog` is a string
+  console.log( this[ key ] );
 
+}, obj ); // <-- the last arg is `thisArg`
 
-    // 6.A.3.2
-    // Naming arrays
+// Prints...
 
-    `dogs` is an array of `dog` strings
+// 'foo'
+// 'bar'
+// 'qux'
+```
 
+`thisArg` can be used with `Array.prototype.every`, `Array.prototype.forEach`, `Array.prototype.some`, `Array.prototype.map`, `Array.prototype.filter`.
 
-    // 6.A.3.3
-    // Naming functions, objects, instances, etc
+### 2.7.3. Binding `this` to the function scope
 
-    camelCase; function and var declarations
+Beyond the generally well known use cases of `call` and `apply`, always prefer `.bind( this ) / CKEDITOR.tools.bind( fn, this )` for creating `BoundFunction` definitions for later invocation.
 
+```javascript
+function Device( opts ) {
+	stream.read( opts.path, CKEDITOR.tools.bind( function( value ) {
+		this.value = value;
+	}, this ) );
+}
+```
 
-    // 6.A.3.4
-    // Naming constructors, prototypes, etc.
+Only resort to aliasing when no preferable option is available.
 
-    // constructor function
-    PascalCase;
-    // but camel case when making public:
-    CKEDITOR.sth.camelCase;
+```javascript
+function Device( opts ) {
+	var self = this;
+	stream.read( opts.path, function( value ) {
+		self.value = value;
+	} );
+}
+```
 
+## 2.8. Misc
 
-    // 6.A.3.6
-    // From the Google Closure Library Style Guide
+This section will serve to illustrate ideas and concepts that should not be considered dogma, but instead exists to encourage questioning practices in an attempt to find better ways to do common JavaScript programming tasks.
 
-    functionNamesLikeThis;
-    variableNamesLikeThis;
-    ConstructorNamesLikeThis;
-    EnumNamesLikeThis;
-    methodNamesLikeThis;
-    SYMBOLIC_CONSTANTS_LIKE_THIS;
+### 2.8.1. Early returns
 
-    ```
+Early returns promote code readability with negligible performance difference.
+```javascript
+// Bad:
+function returnLate( foo ) {
+	var ret;
 
-    B. Faces of `this`
+	if ( foo ) {
+		ret = 'foo';
+	} else {
+		ret = 'quux';
+	}
+	return ret;
+}
```

Comment:
This example is unrealistic (I know it was simplified to be easy to understand) but looking on it, you think "why not go with a one-liner `return foo ? 'foo' : 'quux'`;". And yes, it promotes bad code as a side effect.

Reply:
An example from our code:
https://github.com/ckeditor/ckeditor4/blob/294ea3333aa3d2b796f27ca8503ac7c1a658d50a/plugins/tableselection/plugin.js#L862-L890